lighthouse-core/lib/stack-collector.js

```diff
@@ -15,45 +15,51 @@
 
 const fs = require('fs');
 const libDetectorSource = fs.readFileSync(
-  require.resolve('js-library-detector/library/libraries.js'),
-  'utf8'
-);
+  require.resolve('js-library-detector/library/libraries.js'), 'utf8');
+
+/**
+ * @typedef JSLibraryDetectorTest
+ * @property {string} icon Essentially an id, useful if no npm name is detected.
+ * @property {string} url
+ * @property {string|null} npm npm module name, if applicable to library.
+ * @property {function(Window): false|{version: string|null} | Promise<false|{version: string|null}>} test Returns false if library is not present, otherwise returns an object that contains the library version (set to null if the version is not detected).
+ */
 
 /**
  * @typedef JSLibrary
  * @property {string} name
- * @property {string} version
- * @property {string} npm
- * @property {string} iconName
+ * @property {string} icon
+ * @property {string|null} version
+ * @property {string|null} npm
  */
 
 /**
  * Obtains a list of detected JS libraries and their versions.
  */
 /* istanbul ignore next */
-function detectLibraries() {
+async function detectLibraries() {
   /** @type {JSLibrary[]} */
   const libraries = [];
 
   // d41d8cd98f00b204e9800998ecf8427e_ is a consistent prefix used by the detect libraries
   // see https://github.com/HTTPArchive/httparchive/issues/77#issuecomment-291320900
+  /** @type {Record<string, JSLibraryDetectorTest>} */
   // @ts-ignore - injected libDetectorSource var
-  Object.entries(d41d8cd98f00b204e9800998ecf8427e_LibraryDetectorTests).forEach(
-    async ([name, lib]) => {
-      // eslint-disable-line max-len
-      try {
-        const result = await lib.test(window);
-        if (result) {
-          libraries.push({
-            name: name,
-            version: result.version,
-            npm: lib.npm,
-            iconName: lib.icon,
-          });
-        }
-      } catch (e) {}
-    }
-  );
+  const libraryDetectorTests = d41d8cd98f00b204e9800998ecf8427e_LibraryDetectorTests; // eslint-disable-line 
+
+  for await (const [name, lib] of Object.entries(libraryDetectorTests)) {
```

Comment:
do we get anything out of this? I assumed nearly all of them are synchronous, if they make network requests or something that wouldn't be good 😨

Reply:
> do we get anything out of this

well, first my change makes no sense...`Object.entries` doesn't return an async iterator :) We really only need an `async` on `detectLibraries` and to ditch the inner function calls.

We do get something out of being async. I believe only the `workbox` test is async, but since the function isn't currently async and waiting for the test to return, we don't detect workbox on test pages (though that fact is currently only available in the hidden `js-libraries` audit).

> if they make network requests or something that wouldn't be good

they don't *today*, and hopefully they won't in the future. The workbox one is async because it's [going through the SW APIs to get the SW source file](https://github.com/johnmichel/Library-Detector-for-Chrome/blob/68a527631db0a256932b001e17e8e173e591b55a/library/libraries.js#L1400).

Reply:
ah ok gotcha 👍